### PR TITLE
Support fetching beakerlib libraries in discover

### PR DESCRIPTION
**Pull request by psss**
Initial implementation for beakerlib libraries during discover.
Supports backward-compatbile 'library(component/lib)' format as
well as specifying beakerlib library using a full fmf identifier.
Extend the L1 'require' attribute specification with fmf id.
Update the examples page with a new Test Libraries section.
Include test, confirms that libs requiring libs work as well.
Modify tmt.utils.format to prevent crash on list of dictionaries.

### CODE REVIEW

**Comment by psss** [comment]
Example output of the discover step:

```
/plan/apache
summary: Run only the apache test
    discover
        workdir: /var/tmp/tmt/run-046/plan/apache/discover
        how: fmf
        directory: /home/psss/git/tmt
            Copy '/home/psss/git/tmt' to '/var/tmp/tmt/run-046/plan/apache/discover/default/tests'.
        path: tests/discover/libraries/data
        names: apache
            Check metadata tree in '/var/tmp/tmt/run-046/plan/apache/discover/default/tests/tests/discover/libraries/data'.
            Fetch beakerlib library 'httpd/http'.
            Fetch beakerlib library 'openssl/certgen'.
        summary: 1 test selected
            /apache
```

@sopos, please have a look. As for now libraries are stored under `libs` in the discover step workdir.

**Comment by sopos** [comment]
The the library it self is not represented by some virtual object (test case)? Meaning the url need to be hardcoded? As I've assumed in https://github.com/beakerlib/beakerlib/issues/7, the object would be created in the plan where all the references would be defined by the plan maintainer, thus more flexible. On the other hand if he library maintainer changes location or something related the library maintainer would need to reflect such changes.
All in all, it is still hard to say what solution is the best one as the tests referenced in the plan might be maintained by others so the list of the dependencies may vary. The only logical solution is to have some central solution with an auto resolution. But there's still a question whether to group by the component or not.

**Comment by sopos** [comment]
What about a central registry as a special plan which would live at agreed place where all the libraries would have their own virtual test cases with their `provides` which would be used for the dependency resolution and held the references to the actual place?

**Comment by lukaszachy** [comment]
I need to get over breaking require definition (used to be `whatever dnf install accepts`), but I see why it is there.
However can 'fmf id' way be more generic? E.g. to allow install arbitrary fmf-id test as dependency?
E.g. `fmf-ref{fmf-id}` or similar instead of `library(name)`

We have `recommend` attribute as well - while I don't think that library will be in recommend we should support that. Also libraries will have their `recommend` for sure (to workaround unavailable packages)

**Comment by thrix** [comment]
I like the proposal as the first tangible version @psss ! Just one note, I would like to see `require` support dictionary so we can describe `library{fmf_id}' as a dictionary instead, this string looks weird to me in our nice fmf / yaml format :) 

so instead of:

```
requires:
- openssh
- 'library{url: "https://github.com/psss/fmf", "ref": 0.10}'
```

I would like to rather see:
```
requires:
- openssh
- library:
  url: "https://github.com/psss/fmf"
  ref: 0.1.0
```
path: /examples/wget
name: /download/test

**Comment by sopos** [comment]
Referencing a concrete url and commit for the library is definitely a nice possibility but as I stated earlier, and @lukaszachy requests basically the same, I would like to see the reference to a virtual object which definition would be somewhere (configurable with some reasonable default) so it can be changed in a different environment.

**Comment by kkaarreell** [comment]
May I ask why "library" is used in the reference? I understand that we have beakerlib libraries but I guess that such reference could be used for any fmf (test) object, right? Is 'library' the appropriate word for such use case? I am not saying it is not, seems acceptable given that we are not running it directly but rather only ensuring it is in place. I am just curious whether alternative names were also considered?

**Comment by psss** [comment]
The latest proposal is to support `library(repo/lib)` for backward compatibility and a dictionary format with fmf identifier and specified type. Here's example of the second:

```yaml
require:
    - regular-rpm-package
    - type: beakerlib-library
      url: https://github.com/beakerlib-libraries/openssl
      ref: master
      name: certgen
```

I'm working on refreshing the pull request based on yesterday's feedback.

**Comment by kkaarreell** [comment]
Why did you refer to the type as 'beakerlib-library'? What makes the feature so beakerlib specific that it can't/shouldn't be use by non-beakerlib tests?

**Comment by sopos** [comment]
I also thing that type `library` together with how `beakerlib` is enough. But I might not see something else.

**Comment by lukaszachy** [comment]
> Why did you refer to the type as 'beakerlib-library'? What makes the feature so beakerlib specific that it can't/shouldn't be use by non-beakerlib tests?

All 'rlXXX' statements require beakerlib, usually you need to run them within existing journal.

**Comment by lukaszachy** [comment]
> I also thing that type `library` together with how `beakerlib` is enough. But I might not see something else.

`beaker-library` requires placement on path where `rlImport` finds it. What if we end up with arbitrary libraries? E.g. pytest fixtures... Those "libraries" will have their own requirements on placement, etc.

**Comment by kkaarreell** [comment]
Understood. Thank you for the clarification. I assume that location would be hardcoded for now, right? Btw, what about library persistency? Any considered options to alter that?

**Comment by psss** [comment]
Thinking a bit more about @kkaarreell's comment: The fmf id part of the implementation really is nothing more than clone given repository into the desired subdirectory under `libs` and check it's fmf metadata for possible `require` and `recommend`. So one could easily reference his/her git repo with shared (non-beakerlib) stuff and just include the code for example using `../../../../libs/library/lib.sh` as the tests are doing now, until it's implemented in `rlImport`. So what is actually the point of the `type` key? Should we have rather `destination` which would default to `libs`?

**Comment by psss** [comment]
Implemented what was agreed on last week and rebased on the latest master. @lukaszachy, @thrix, @sopos, @kkaarreell, ready for review.

File: tmt/steps/discover/fmf.py [hunk 1]
```diff
@@ -82,6 +90,72 @@ def wake(self):
...
+            self.run(['git', 'clone', url, directory], shell=False)
+            self.run(['git', 'checkout', ref], shell=False, cwd=directory)
+
+        # Check library metadata for requires
+        tree = fmf.Tree(directory)
+        library = tree.find(f"/{name}")
+        requires = tmt.utils.listify(library.get('require', []))
+        return self.check_library_require(requires)
+
+    def check_library_require(self, requires):
```
[the remark this answers is not in the document]
Comment (psss) [reply tmt/steps/discover/fmf.py]: Good point. Implemented.

File: tests/discover/libraries/data/apache.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+# vim: dict+=/usr/share/beakerlib/dictionary.vim cpt=.,w,b,u,t,i,k
+. /usr/share/beakerlib/beakerlib.sh || exit 1
+
+rlJournalStart
+    rlPhaseStartSetup
+        rlAssertRpm "httpd"
+        #rlRun "rlImport httpd/http"
```
Comment (lukaszachy) [inline tests/discover/libraries/data/apache.sh:8]: If we can't use rlImport yet, could you mark it with TODO or FIXME? 
While it is great that `source ../../` works _the_ supported way is rlImport and IMHO we should test that

File: tmt/steps/discover/fmf.py [hunk 3]
```diff
@@ -82,6 +185,31 @@ def wake(self):
             if value:
                 self.data[option] = value
 
+    def check_requires(self, all_requires):
```
Comment (lukaszachy) [inline tmt/steps/discover/fmf.py:188]: IMO misleading - How about `check_dependencies`? And even better if 'check' can be renamed - it just detects those, no checking at all

File: tmt/steps/discover/fmf.py [hunk 4]
```diff
@@ -82,6 +185,31 @@ def wake(self):
...
+        """
+        Check requires for possible beakerlib libraries
```
Comment (lukaszachy) [inline tmt/steps/discover/fmf.py:190]: requires and recommends, or dependencies to say it shortly.

File: tmt/steps/discover/fmf.py [hunk 5]
```diff
@@ -82,6 +185,31 @@ def wake(self):
...
+        """
+        requires = []
+        recommends = []
+        for require in all_requires:
+            # Library require
+            try:
+                library = Library(require, parent=self)
+                recommends.extend(library.recommend)
+                # Recursively check for possible dependent libraries
+                require, recommend = self.check_requires(library.require)
```
Comment (lukaszachy) [inline tmt/steps/discover/fmf.py:205]: Note: Should we explicitly document in recommend field that libraries are not supported there and user has to use require attribute?

File: tests/discover/libraries/data/apache.sh (hunk 2 above)
Comment (psss) [reply tests/discover/libraries/data/apache.sh:8]: Makes sense added.

File: tmt/steps/discover/fmf.py (hunk 3 above)
Comment (psss) [reply tmt/steps/discover/fmf.py:188]: Yes, that method name is a challenge. During implementation I have renamed it already three times ;-) Let's call it just `dependencies` ;-)

File: tmt/steps/discover/fmf.py (hunk 4 above)
Comment (psss) [reply tmt/steps/discover/fmf.py:190]: Updated.

File: tmt/steps/discover/fmf.py (hunk 5 above)
Comment (psss) [reply tmt/steps/discover/fmf.py:205]: There is never enough explicit documentation. Added :)

**Comment by psss** [comment]
@lukaszachy, thanks for the feedback. All your comments should be addressed. I've also modified tests to use container except for the apache test which needs a full virtualization support.

**Comment by psss** [comment]
Thanks for the review, @lukaszachy. I've fixed wrong test path and I'm going to merge the thing.

**Pull request merged**